(Feature) Render Server HTML, PDF

sfount opened the pull request:
This pull request implements server side html and pdf rendering. It uses the [handlebars](http://handlebarsjs.com) templating library to allow clean templates that are similar to those found on the client ( {{}} syntax ). 

It also:
- Includes `ng-sanitize` to allow rendered HTML to be included in modules, as well as trusting resources URLs used in PDF embedding
- Implements the PDF renderer, using it for the patient invoice receipt success state. 
- Removes legacy reports on the server - favouring a structure that encourages importing required renderers and removing duplicated resources files (en.json/ fr.json etc.) 
- Researches methods of printing PDFs that have been downloaded from the server without forcing the user to use the embedded object

**json renderer**
![json](https://cloud.githubusercontent.com/assets/2844572/15392449/14d620a8-1dbf-11e6-91e1-ba627cb4149c.png)

**html renderer**
![html](https://cloud.githubusercontent.com/assets/2844572/15392456/1bf12ab8-1dbf-11e6-81c0-f43dee222472.png)

**pdf renderer**
![pdf](https://cloud.githubusercontent.com/assets/2844572/15392461/2254fba0-1dbf-11e6-96f3-61a5e93474b8.png)

**Use case: Patient Invoice Receipt**
![receipt_pdf](https://cloud.githubusercontent.com/assets/2844572/15392468/2f5f807c-1dbf-11e6-890f-22e95d12bfeb.png)

This method allows us to create reports that are always driven by the same underlying data but can be presented to the user in any number of ways depending on requirement (printing etc.). It also considers the user experience, ensuring the user is not required to leave the application to retrieve a PDF.

This pull request builds on previous work found https://github.com/IMA-WorldHealth/bhima-2.X/pull/260 and https://github.com/IMA-WorldHealth/bhima-2.X/pull/293

---
- [x] Run your code through JSHint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
- [x] Run integration tests.
- [x] Run end-to-end tests.
- [x] Accurately described the changes your are making in this pull request.

sfount committed:
(Feat.) Introduce core html render; update depend.

This commit adds the core html rendering structure that can be used for
receipts/ reports. The current architecture is set up to encourage
templating and rendering on the server with this information passed to
the client. This should be reviewed and discussed.

* Added angular sanitize as a dependency for the client - this allows
  HTML to be dynamically templated into a module (in this case a
  receipt).
* HTML render case in the receipt modal component
* Express handlebars dependency added to allow templates to be defined
  that are familiar to those using the current technology stack (angular
  templates)
* Defined initial pattern for how to translate/ filter currency on the
  the server using handlebar `helpers`, this should be reviewed and
  improved upon with a express handlebars wrapper to set up the instance
  and define all required helpers.

sfount committed:
(Feat.) Core PDF renderer, printing on the client

This commit introduces the basic PDF renderer for generating server side
documents, the PDF renderer imports and makes use of the `html` renderer
introduced in the previous commit and passes this through `wkhtmltopdf`.

* Added stream to promise library to ensure that the current renderer
API can be respected by `wkhtmltopdf`. Waiting on the entire stream to
be complete means not allowing the client to receive 'chunks' of the PDF
as it is available however this issue can be addressed if it becomes a
problem.
* Updated renderer API to allow individual renderers to set their own
headers
* Client side `iframe` implemented to embed PDF documents - this allows
a separate environment that can be specifically invoked to print

sfount committed:
(Chore) Clean up server side templating process

This commit re-factors the use of the express handlebars library into a wrapper file `template.js` that allows configuring helper methods only once. It also cleans up comments and left over console logs.

sfount committed:
(Chore) Re-factor client side pdf receipts

This commit cleans up comments and services used for client side presentation of reports. Temporary PDF code is moved into PDF specific blocks and only run given the relevant renderer.
Removed legacy client side receipt that was rendered using the 'json' route.

sfount committed:
(Chore) Remove 1.x report files and references

This commit deletes the `server/controllers/reports` folder that uses an outdated infrastructure of internally routing logic. All previous dot templates and context files have also been removed. Updated routes.js references.

sfount committed:
(Fix) Patient Invoice module completion state (PDF Receipts)

This commit cleans up the patient invoice module in order to integrate PDF receipts as the completion state.
 * Introduces temporary success notification sign on the receipt modal
 * Adds client side validation to patient invoice notes in order to not make frequent bad requests
 * Uses the Notify service on invalid invoice items and form fields
 * Resolves a bug in the patient date component that always shows invalid state on submission
 * Ensures the invoice module can be cleared to allow the receipt to complete the module state. Closes https://github.com/IMA-WorldHealth/bhima-2.X/issues/286

sfount committed:
(Chore) Updated invoice E2E for PDF receipt modal

This commit updates what should be expected in e2e tests on completion of the sale module. The actual invoice submission tests are still skipped due to a bug in the posting process. This was recorded here https://github.com/IMA-WorldHealth/bhima-2.X/issues/405.

* Fixed minor issue with `notes` form validation

sfount committed:
Add test cases for additional PDF/HTML/JSON renderers

* Additional test cases expecting content-types and non empty bodies
* Cleaned up console logs/ legacy tests on server rendering

jniles commented on client/src/js/services/receipts/ReceiptModal.js:
```diff
@@ -2,52 +2,55 @@ angular.module('bhima.services')
+   * Invokes a patient invoice receipt
+   * 
+   * @param {String} uuid             Target invoice UUID
+   * @param {Boolean} notifyCreated   Defines if a success message should be shown for entity creation
```
Nice docs!

jniles commented on client/src/js/services/receipts/ReceiptService.js:
```diff
@@ -21,14 +19,19 @@ function ReceiptService($http, util) {
-   * @param {String} uuid      Target invoice UUID to report on
-   * @param {Object} options   Configuration options for the server generated
+   * @params {String} uuid      Target invoice UUID to report on
+   * @params {Object} options   Configuration options for the server generated
```
The correct format for this tag is [@param](http://usejsdoc.org/tags-param.html), not `@params`.

jniles commented on client/src/js/services/receipts/modal/receiptModalController.js:
```diff
@@ -10,31 +10,49 @@ ReceiptModalController.$inject = ['$uibModalInstance', '$window', 'receipt', 'op
+ * @params {String} render    Render target used to generate report
  */
-function ReceiptModalController($modalInstance, $window, receipt, options) {
+function ReceiptModalController($modalInstance, $window, $sce, Notify, receipt, options) {
```
Do you still need the NotifyService here?

jniles commented on client/src/js/services/receipts/modal/receiptModalController.js:
```diff
@@ -10,31 +10,49 @@ ReceiptModalController.$inject = ['$uibModalInstance', '$window', 'receipt', 'op
+  function print() {
+    /**@todo This printing could be exposed by a directive/ component */
+    // special case for embedded content
+    if (options.renderer === 'pdf') {
```
There are several places where this check occurs.  I wonder if it might be better to have a constant (or service) like `renderers.PDF` that these checks could use so that it isn't a hardcoded string all the time.  something like:

``` js
if (options.renderer === renderers.PDF) {
  // logic
}
```

sfount replied:
Good idea. :+1:

jniles commented on client/src/js/services/receipts/modal/receiptModalController.js:
```diff
@@ -10,31 +10,49 @@ ReceiptModalController.$inject = ['$uibModalInstance', '$window', 'receipt', 'op
+    /**@todo This printing could be exposed by a directive/ component */
+    // special case for embedded content
+    if (options.renderer === 'pdf') {
+      $window.frames.pdf.contentWindow.print();
```
This is fascinating.  Just for my information, is the `frames.pdf` named that way because the iframe's `id` is 'pdf'?

sfount replied:
Yes. Ideally this should be commented. 

As the `@todo` mentions said this could be encapsulated in a component that exposes an API however I think would benefit from a separate body of work (issue, design, unit tests etc.).

jniles replied:
Yeah, something this specific isn't going to be consumed by a ton of developers.  This is fine, but I was curious.

jniles commented on client/src/partials/patient_invoice/patientInvoice.js:
```diff
@@ -105,8 +114,9 @@ function PatientInvoiceController($q, $location, Patients, PriceLists, PatientIn
   // reset everything in the controller - default values
-  function clear() {
-
+  function clear(detailsForm) {
```
~~`detailsForm` is an unused parameter.~~  EDIT, found where it was used.  Sorry sorry.

jniles commented on client/src/partials/templates/bhDateEditor.tmpl.html:
```diff
@@ -10,9 +10,10 @@
   <div
     class="form-group"
-    ng-class="{ 'has-error' : $ctrl.validationTrigger && DateEditorForm.$error } "
+    ng-class="{ 'has-error' : $ctrl.validationTrigger && DateEditorForm.$invalid } "
```
Nice catch!

jniles commented on package.json:
```diff
@@ -46,8 +47,10 @@
     "node-uuid": "^1.4.7",
     "numeral": "^1.5.3",
     "q": "~1.4.1",
+    "session-file-store": "0.2.0",
```
I think this was accidentally added back in.  Could you remove session file store?

jniles commented on server/controllers/finance/reports/invoice.receipt.handlebars:
```diff
@@ -0,0 +1,73 @@
+<head>
+  <title>{{translate "TREE.PATIENT_INVOICE"}} {{reference}}</title>
```
This is super cool!

jniles commented on server/lib/renderers/pdf.js:
```diff
@@ -0,0 +1,45 @@
+const wkhtmltopdf     = require('wkhtmltopdf');
+const q               = require('q');
+const streamToPromise = require('stream-to-promise');
+const process         = require('process');
```
Didn't realize you could `require()` process.  Interesting!  Could you list it in the `@requires` block?

jniles commented on server/test/api/invoiceReceipt.js:
```diff
@@ -13,12 +13,14 @@ describe('/reports/invoices Receipts Interface', function () {
   const invoiceItemLength   = 1;
   const invalidInvoice      = 'unknown';
-  const validRenderTarget   = '?render=json';
+  const jsonRenderTarget   = '?render=json';
```
Just FYI, `superagent` (what `chai-http` is based on) accepts query parameters as an object, similar to angular.  I think it may be possible to do it with `chai-http`, if you think it looks nicer.  [Here is the relevant documentation](https://visionmedia.github.io/superagent/#query-strings).

jniles commented on client/test/e2e/patient/invoice.spec.js:
```diff
@@ -99,7 +100,7 @@ describe('Patient Invoice', function () {
   });
 
-  it.skip('blocks submission for an invalid grid', function () {
+  it('blocks submission for an invalid grid', function () {
```
Yay!

sfount committed:
(fix) Add `wkhtmltopdf` dependency to travis build

This commit apt-get install wkhtmltopdf to ensure the test runner has
access to creating PDF files. It also addresses a number of changes raised
during code review and resolves an issue with end to end test ordering.

jniles commented on server/lib/renderers/html.js:
```diff
@@ -0,0 +1,22 @@
+ * @requires express-handlebars
+ * @requires server/lib/template
+ */
+const exhbs = require('express-handlebars');
```
If you put 'use strict' above this line with `const`, Travis will pass.

For Node versions 5 and under, `let`, `const`, and others have to be within 'strict' mode.  Node 6 removes this restriction.

Unfortunately, our error stack traces only went as far as `invoice.js` which requires this file.  Perhaps we should look into handlers that give us long stack traces.

jniles commented on server/lib/renderers/pdf.js:
```diff
@@ -0,0 +1,46 @@
+ * @requires q
+ * @requires process
+ */
+const wkhtmltopdf     = require('wkhtmltopdf');
```
This line will also need a 'use strict' directive above it.

jniles commented on server/lib/template.js:
```diff
@@ -0,0 +1,76 @@
+ * 
+ * @returns {Function} handlebars render method, accepting a template and a context
+*/
+const exphbs = require('express-handlebars');
```
A final 'use strict' directive should be placed above this line.
